fence/agents/powerman/fence_powerman.py

```diff
@@ -0,0 +1,300 @@
+			logging.warning("command '%s' returned (%s) but query reported (%s) afterwards\n", cmd, result, queryret)
+
+		if queryret == "on":
+			return 0
+		else:
+			return -1
+
+	def off(self, host):
+		logging.debug("PowerMan off: %s\n", host)
+		cmd = ["--off", host]
+		try:
+			result, ret_code = self._run(cmd, True)
+		except OSError as e:
+			logging.error("PowerMan Error: The command '%s' failed: %s\n", cmd, e)
+			return -1
+		except ValueError as e:
+			logging.error("PowerMan Error: Popen: invalid arguments: %s\n", e)
+			return -1
+		logging.debug("pm.off result: %s ret_code: %s\n", result, ret_code)
+
+		## For some devices, the "on" or "off" command can fail silently,
+		## but for query powerman keeps trying until it gets a valid answer;
+		## so we must check after trying to power on/off a device.
+		## Powerman seems to report the old state if asked too quickly, wait a second.
+
+		time.sleep(1)
+		queryret = self.query(host)
+
+		if (ret_code == 0 and queryret == "on") or (ret_code < 0 and queryret == "off"):
+			logging.warning("command '%s' returned (%s) but query reported (%s) afterwards\n", cmd, result, queryret)
+
+		if queryret == "off":
+			return 0
+		else:
+			return -1
+
+	def list(self):
+		## Error checking here is faulty.  Try passing
+		## invalid args, e.g. --query --exprange to see failure
+		cmd = ["-q","--exprange"]
+		try:
+			result, ret_code = self._run(cmd, False)
+		except OSError as e:
+			logging.error("PowerMan Error: The command '%s' failed: %s\n", cmd, e)
+			return -1
+		except ValueError as e:
+			logging.error("PowerMan Error: Popen: invalid arguments: %s\n", e)
+			return -1
+		if ret_code < 0:
+			# there was an error with the command
+			return ret_code
+		else:
+			state = {}
+			for line in result[0].split('\n'):
+				if len(line) > 2:
+					fields = line.split(':')
+					if len(fields) == 2:
+						state[fields[0]] = (fields[0],fields[1])
+			return state
+
+	def query(self, host):
+		cmd = ["--query", host]
+		try:
+			result, ret_code = self._run(cmd, True)
+		except OSError as e:
+			logging.error("PowerMan Error: The command '%s' failed: %s\n", cmd, e)
+			return -1
+		except ValueError as e:
+			logging.error("PowerMan Error: Popen: invalid arguments: %s\n", e)
+			return -1
+		if ret_code < 0:
+			# there was an error with the command
+			return ret_code
+		else:
+			res = result.split('\n')
+			res = [r.split() for r in res]
+			# find the host in command's returned output
+			for lst in res:
+				if lst[0] == 'No' and lst[1] == 'such' and lst[2] == 'nodes:':
+					return -1
+				if host in lst:
+					return lst[0][:-1] # lst[0] would be 'off:'-- this removes the colon
+			# host isn't in the output
+			return -1
+
+
+def get_power_status(conn, options):
+	logging.debug("get_power_status function:\noptions: %s\n", str(options))
+	pm = PowerMan(options['--ip'], options['--ipport'])
+	# if Pacemaker is checking the status of the Powerman server...
+	if options['--action'] == 'monitor':
```

Comment:
Monitor action tests if fence device is running what you do test correctly. But 'response' off does not mean that fence device is off but it means that computer is off -> so monitor action will suceed. You have to return something else e.g. 'error' when PM is not running